File: common/config/raft/store.go
/*
 * Copyright (c) 2019-2021. Abstrium SAS <team (at) pydio.com>
 * This file is part of Pydio Cells.
 *
 * Pydio Cells is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Pydio Cells is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with Pydio Cells.  If not, see <http://www.gnu.org/licenses/>.
 *
 * The latest code can be found at <https://pydio.com>.
 */

package raft

import (
	"bytes"
	"encoding/gob"
	"encoding/json"
	"log"
	"sync"

	"github.com/pydio/cells/x/configx"

	"go.etcd.io/etcd/etcdserver/api/snap"
	"go.etcd.io/etcd/raft/raftpb"
)

// a key-value store backed by raft
type KVStore struct {
	proposeC    chan<- string // channel for proposing updates
	mu          sync.RWMutex
	kvStore     configx.Values // current committed key-value pairs
	snapshotter *snap.Snapshotter
}

type KV struct {
	Key string
	Val interface{}
}

func init() {
	gob.Register(map[string]interface{}{})
}

func NewKVStore(store configx.Values, snapshotter *snap.Snapshotter, proposeC chan<- string, commitC <-chan *commit, errorC <-chan error) *KVStore {
	s := &KVStore{proposeC: proposeC, kvStore: store, snapshotter: snapshotter}
	snapshot, err := s.loadSnapshot()
	if err != nil {
		log.Panic(err)
	}
	if snapshot != nil {
		log.Printf("loading snapshot at term %d and index %d", snapshot.Metadata.Term, snapshot.Metadata.Index)
		if err := s.recoverFromSnapshot(snapshot.Data); err != nil {
			log.Panic(err)
		}
	}
	// read commits from raft into kvStore map until error
	go s.readCommits(commitC, errorC)
	return s
}

func (s *KVStore) Lookup(key string) (string, bool) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	v := s.kvStore.Val(key)
	if v == nil {
		return "", false
	}
	return v.Default("").String(), true
}

func (s *KVStore) Propose(k string, v interface{}) {
	var buf bytes.Buffer
	if err := gob.NewEncoder(&buf).Encode(KV{k, v}); err != nil {
		log.Fatal(err)
	}
	s.proposeC <- buf.String()
}

func (s *KVStore) readCommits(commitC <-chan *commit, errorC <-chan error) {
	for commit := range commitC {
		if commit == nil {
			// signaled to load snapshot
			snapshot, err := s.loadSnapshot()
			if err != nil {
				log.Panic(err)
			}
			if snapshot != nil {
				log.Printf("loading snapshot at term %d and index %d", snapshot.Metadata.Term, snapshot.Metadata.Index)
				if err := s.recoverFromSnapshot(snapshot.Data); err != nil {
					log.Panic(err)
				}
			}
			continue
		}

		for _, data := range commit.data {
			var dataKv KV
			dec := gob.NewDecoder(bytes.NewBufferString(data))
			if err := dec.Decode(&dataKv); err != nil {
				log.Fatalf("raftexample: could not decode message (%v)", err)
			}
			s.mu.Lock()
			s.kvStore.Val(dataKv.Key).Set(dataKv.Val)
			s.mu.Unlock()
		}
		close(commit.applyDoneC)
	}
	if err, ok := <-errorC; ok {
		log.Fatal(err)
	}
}

func (s *KVStore) GetSnapshot() ([]byte, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return json.Marshal(s.kvStore)
}

func (s *KVStore) loadSnapshot() (*raftpb.Snapshot, error) {
	snapshot, err := s.snapshotter.Load()
	if err == snap.ErrNoSnapshot {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	return snapshot, nil
}

func (s *KVStore) recoverFromSnapshot(snapshot []byte) error {
	var store map[string]string
	if err := json.Unmarshal(snapshot, &store); err != nil {
		return err
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	s.kvStore.Val().Set(store)
	return nil
}

File: frontend/front-srv/assets/gui.ajax/res/js/ui/Components/dynamic-grid/Store.js
/*
 * Copyright 2007-2017 Charles du Jeu - Abstrium SAS <team (at) pyd.io>
 * This file is part of Pydio.
 *
 * Pydio is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Pydio is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with Pydio.  If not, see <http://www.gnu.org/licenses/>.
 *
 * The latest code can be found at <https://pydio.com>.
 */

class Store extends Observable{

    /**
     * Init a card store
     * @param prefNamespace Namespace for getting/setting user preferences
     * @param defaultCards Array of cards to be displayed by default
     */
    constructor(prefNamespace, defaultCards, pydioObject){
        super();
        this._namespace = prefNamespace;
        this._pydio = pydioObject;
        this._cards = this.getUserPreference("Cards");
        if(!this._cards){
            this._cards = defaultCards;
        }
    }

    getUserPreference(prefName){
        return this._pydio.user.getLayoutPreference(this._namespace + prefName, null)
    }

    saveUserPreference(prefName, prefValue){
        this._pydio.user.setLayoutPreference(this._namespace + prefName, prefValue)
    }

    saveCards(cards){
        this.saveUserPreference('Cards', cards);
    }

    resetCards(){
        this.saveUserPreference('Cards', null);
    }

    setCards(newCards){
        this._cards = newCards;
        this.notify("cards", this._cards);
        this.saveCards(newCards);
    }

    getCards(){
        return this._cards;
    }

    removeCard(cardId){
        var index = -1;
        var currentCards = this.getCards();
        currentCards.map(function(card, arrayIndex){
            if(card.id == cardId) index = arrayIndex;
        });
        if(index == -1){
            console.warn('Card ID not found, this is strange.', cardId);
            return;
        }
        var newCards;
        if(index == 0) newCards = currentCards.slice(1);
        else if(index == currentCards.length-1) newCards = currentCards.slice(0, -1);
        else newCards = currentCards.slice(0,index).concat(currentCards.slice(index+1));
        this.setCards(newCards);
    }

    createCardId(cardDefinition, randomize=false){
        var id = LangUtils.computeStringSlug(cardDefinition['title']);
        if(randomize){
            id += '-' + Math.round(Math.random() * 100 + 10);
        }
        var alreadyExists = false;
        this._cards.map(function(card){
            if(card.id == id) alreadyExists = true;
        }.bind(this));
        if(alreadyExists){
            id = this.createCardId(cardDefinition, true);
        }
        return id;
    }

    addCard(cardDefinition){
        //console.log(cardDefinition);

        cardDefinition['id'] = this.createCardId(cardDefinition);
        this.setCards(this._cards.concat([cardDefinition]));
    }

}

export {Store as default}